lib/BloqadeQMC/src/updates/cluster.jl

```diff
@@ -0,0 +1,333 @@
+            ccount += 1
+            push!(cstack, i)
+            in_cluster[i] = ccount
+            if qmc_state isa BinaryGroundState && !(qmc_state.trialstate isa AbstractProductState)
+                left_flips = empty!(qmc_state.trialstate.left_flips)
+                right_flips = empty!(qmc_state.trialstate.right_flips)
+            end
+
+            empty!(current_cluster)
+            push!(current_cluster, i)
+            lnA = 0.0
+            if qmc_state isa BinaryGroundState
+                lnA += trialstate_weight_change(qmc_state, lsize, Ns, i)
+            end
+
+            while !isempty(cstack)
+                leg = LinkList[pop!(cstack)]
+
+                if iszero(in_cluster[leg])
+                    in_cluster[leg] = ccount  # add the new leg and flip it
+                    push!(current_cluster, leg)
+                    if qmc_state isa BinaryGroundState
+                        lnA += trialstate_weight_change(qmc_state, lsize, Ns, i)
+                    end
+                    a = Associates[leg]
+
+                    a == 0 && continue
+                    # from this point on, we know we're on a bond op
+                    op = operator_list[op_indices[leg]]
+                    w = getweightindex(H, op) - getoperatortype(H, op)
+                    preflip_bond_type, postflip_bond_type = update_kernel!(qmc_state, H, ccount, leg, a)
+                    lnA += (
+                        getlogweight(H.op_sampler, w + postflip_bond_type)
+                        - getlogweight(H.op_sampler, w + preflip_bond_type)
+                    )
+                end
+            end
+
+            if qmc_state isa BinaryGroundState && !(qmc_state.trialstate isa AbstractProductState)
+                lnA += logweightchange(qmc_state.trialstate, qmc_state.left_config, left_flips)
+                lnA += logweightchange(qmc_state.trialstate, qmc_state.right_config, right_flips)
+            end
+
+            A = acceptance(H, lnA)
+            fit!(runstats, :cluster_update_accept, A)
+
+            if rand(rng) < A
+                @inbounds for j in current_cluster
+                    LegType[j] ⊻= 1  # spinflip
+                end
+                fit!(runstats, :accepted_cluster_sizes, length(current_cluster))
+                num_accept += 1
+            else
+                fit!(runstats, :rejected_cluster_sizes, length(current_cluster))
+                num_reject += 1
+            end
+            fit!(runstats, :cluster_sizes, length(current_cluster))
+        end
+    end
+
+    if !(runstats isa NoStats)
+        fit!(runstats, :accepted_cluster_count, num_accept+1)
+        fit!(runstats, :rejected_cluster_count, num_reject+1)
+        fit!(runstats, :cluster_count, ccount+1)
+    end
+
+    # map back basis states and operator list
+    ocount = _map_back_basis_states!(rng, lsize, qmc_state, H)
+    _map_back_operator_list!(ocount, qmc_state, H, d)
+
+    return lsize
```

Comment:
It probably worth thinking if this return value can be a state object - it would be nice the MC iteration can roughly follow an iterator interface in Julia. But this is not a blocking comment.

Reply:
Also it seems lsize is the same as input here? If I'm not missing something what's the point of returning the same value?